src/server/pfs/server/api_server.go

```diff
@@ -398,6 +401,101 @@ func (a *apiServer) DeleteBranch(ctx context.Context, request *pfs.DeleteBranchR
 }
 
+// SearchForFileInBranch searches for commits that reference a supplied file being modified in a branch.
+func (a *apiServer) SearchForFileInBranch(ctx context.Context, request *pfs.SearchForFileInBranchRequest) (*pfs.SearchForFileInBranchResponse, error) {
+	timeout, err := types.DurationFromProto(request.Timeout)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	// Use sane defaults for fields that are not defined.
+	if timeout == 0 {
+		timeout = time.Minute * 5
+	}
+	if request.Limit == 0 {
+		request.Limit = 10
+	}
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	var found []*pfs.Commit
+	commit := request.GetStart()
+	commitsSearched := 0
+	returnResp := func(found []*pfs.Commit, commitsSearched int, lastSearchedCommit *pfs.Commit) *pfs.SearchForFileInBranchResponse {
+		resp := &pfs.SearchForFileInBranchResponse{}
+		resp.FoundCommits = found
+		resp.CommitsSearched = int32(commitsSearched)
+		resp.LastSearchedCommit = commit
+		return resp
+	}
+	searchDone := false
+	for {
+		if searchDone {
+			return returnResp(found, commitsSearched, commit), err
+		}
+		if err := log.LogStep(ctx, "searchForFileInCommit", func(ctx context.Context) error {
+			inspectCommitResp, err := a.InspectCommit(ctx, &pfs.InspectCommitRequest{Commit: commit})
+			if err != nil {
+				return err
+			}
+			commit = inspectCommitResp.Commit
+			inCommit, err := a.searchForFileInCommit(ctx, commit, request.FilePath)
+			if err != nil {
+				return err
+			}
+			if inCommit {
+				log.Info(ctx, "found file", zap.String("commit", commit.ID), zap.String("branch", commit.Branch.String()), zap.String("targetFile", request.FilePath))
+				found = append(found, commit)
+			}
+			commitsSearched++
+			if len(found) == int(request.Limit) || inspectCommitResp.ParentCommit == nil {
+				searchDone = true
+				return nil
+			}
+			commit = inspectCommitResp.ParentCommit
+			return nil
+		}, zap.String("commit", commit.ID), zap.String("branch", commit.Branch.String()), zap.String("targetFile", request.FilePath)); err != nil {
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				return returnResp(found, commitsSearched, commit), nil
+			}
+			return nil, errors.EnsureStack(err)
+		}
+	}
+}
+
+func (a *apiServer) searchForFileInCommit(ctx context.Context, commit *pfs.Commit, filePath string) (bool, error) {
+	diffID, err := a.driver.commitStore.GetDiffFileSet(ctx, commit)
+	if err != nil {
+		return false, err
+	}
+	diffFileSet, err := a.driver.storage.Open(ctx, []fileset.ID{*diffID})
+	if err != nil {
+		return false, err
+	}
+	found := false
+	if err = diffFileSet.Iterate(ctx, func(file fileset.File) error {
+		if file.Index().Path == filePath {
+			found = true
+			return errutil.ErrBreak
+		}
+		return nil
+	}); err != nil && !errors.Is(err, errutil.ErrBreak) {
+		return false, err
+	}
+	// a file cannot be in the additive and deletive set, so if its found, skip iteration over deletive.
```

Comment:
Is this true?  I thought you do a put with a delete + add

Reply:
Oh interesting, Is that the case?
Actually this comment should be reworked, because if a file is found in the additive, we can skip iterating over deletive because we don't currently care about the operation, just that the file was found.

Reply:
For the modified case, a match in either would be sufficient.  You can check both in parallel or just one and possibly skip the other.